### PR TITLE
Server wms print layers

### PR DESCRIPTION
**Pull request by mhugent**
This pull request adds the possibility to request external WMS layers (which are not defined in the project) in GetMap and GetPrint requests. In web maps, there  are often WMS/WMTS background layers that we want to have directly in the Web map (not through QGIS server). However, for printing, these layers should be enabled (and WMTS layers replaced by WMS layers). Also, some web clients allow the user to add external WMS services to a web map, which also need to be in the printed output.

The proposed solution is to have the dynamic wms layers as EXTERNAL_WMS:&lt;identifier&gt; in the LAYERS parameter and pass the needed wms parameters as &lt;identifier&gt;:url,&lt;identifier&gt;:format,&lt;identifier&gt;:crs etc. in the request. Like this, it is possible to have several external WMS layers in a request and they can be at any position (not only as background layers). An example to request an external wms layer with a vector layer (defined in the project) is  the following:

http://localhost/cgi-bin/qgis_mapserv.fcgi?SERVICE=WMS&REQUEST=GetMap&LAYERS=EXTERNAL_WMS:sogis_avwms,eisenbahnen&STYLES=,&BBOX=xmin,ymin,xmax,ymax&FORMAT=image/png&sogis_avwms:url=http://geoweb.so.ch/wms/avwms?&sogis_avwms:layers=AVWMS_farbig&sogis_avwms:styles=&sogis_avwms:format=image/png&sogis_avwms:crs=EPSG:21781

Are there any security considerations related to passing external wms services to QGIS server? One thing I could imagine is that someone could try to exploit overflows in image libraries by setting up a wms service that always returns a malicious image. Are there other possible dangers?

If other devs are fine with the proposed approach, I'll add a unit test and I'm planning to extend the GetProjectSettings output such that web clients can easily directly access WMTS layers and replace them by WMS layers for printing.

### CODE REVIEW

**Comment by rldhont** [comment]
@mhugent cool feature
@pblottiere any opinion on code ?

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 1]
```diff
@@ -2633,6 +2651,41 @@ namespace QgsWms
...
+      if ( !it.key().startsWith( externalLayerId.toUpper() + ":" ) )
+      {
+        break;
+      }
+
+      QString paramKey = it.key();
+      paramKey.remove( 0, externalLayerId.size() + 1 );
+      wmsUri.setParam( paramKey.toLower(), it.value() );
+      ++it;
+    }
```
Comment (rldhont) [inline src/server/services/wms/qgswmsrenderer.cpp]: The wmsUri construction has to be part of the QgsWmsParameters class.

**Review by rldhont — CHANGES_REQUESTED** [review]

File: src/server/services/wms/qgswmsrenderer.cpp (hunk 1 above)
Comment (mhugo) [reply src/server/services/wms/qgswmsrenderer.cpp]: @mhugent any answer to @rldhont 's comment ?

Comment (rldhont) [reply src/server/services/wms/qgswmsrenderer.cpp]: @mhugo it has been done

Comment (mhugent) [reply src/server/services/wms/qgswmsrenderer.cpp]: I agree with rldhont's comment and it is adressed in commit https://github.com/qgis/QGIS/pull/5396/commits/c99388804a2a32608fb239fbb29b25a6952f5177

Comment (mhugo) [reply src/server/services/wms/qgswmsrenderer.cpp]: oops, sorry, my bad !

**Pull request merged**